openedx/features/effort_estimation/block_transformers.py

```diff
@@ -0,0 +1,225 @@
+"""
+Effort Estimation Transformer implementation.
+Adds effort estimations for block types it recognizes.
+"""
+
+import math
+
+import crum
+import lxml.html
+from django.utils.functional import cached_property
+from edxval.api import get_videos_for_course
+
+from openedx.core.djangoapps.content.block_structure.transformer import BlockStructureTransformer
+from openedx.core.lib.mobile_utils import is_request_from_mobile_app
+
+from .toggles import EFFORT_ESTIMATION_LOCATION_FLAG
+
+
+class EffortEstimationTransformer(BlockStructureTransformer):
+    """
+    A transformer that adds a partition condition for all graded content
+    so that the content is only visible to verified users.
+
+    This transformer requires that the UserPartitionTransformer also be included in your transformer list.
+    """
+    WRITE_VERSION = 1
+    READ_VERSION = 1
+
+    # Public xblock field names
+    EFFORT_ACTIVITIES = 'effort_activities'
+    EFFORT_TIME = 'effort_time'
+
+    # Private transformer field names
+    DISABLE_ESTIMATION = 'disable_estimation'
+    HTML_WORD_COUNT = 'html_word_count'
+    VIDEO_CLIP_DURATION = 'video_clip_duration'
+    VIDEO_DURATION = 'video_duration'
+
+    CACHE_VIDEO_DURATIONS = 'video.durations'
+    DEFAULT_WPM = 265  # words per minute
+
+    class MissingEstimationData(Exception):
+        pass
+
+    @classmethod
+    def name(cls):
+        """Unique identifier for the transformer's class; same identifier used in setup.py."""
+        return 'effort_estimation'
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Grabs raw estimates for leaf content.
+
+        Pooled estimates higher on verticals and such is done by transform() below, because what each user sees could
+        be different.
+        """
+        block_structure.request_xblock_fields('category')
+        block_structure.request_xblock_fields('global_speed', 'only_on_web')  # video fields
```

Comment:
what is global speed?

Reply:
`global_speed` is the user's preference for the default video speed. `speed` is the last-used speed for a given video. We only look at `global_speed` in this transformer.